[Settings] Adding Tests to Verify Backward Compatibility

laviusmotileng-ms opened the pull request:
## Summary of the Pull Request

- Adding Tests to Verify Backward Compatibility.
- Load Original Settings file, initialize the view model for that module and check if the values initialized in the view model are the same as those loaded from the config file.

## PR Checklist
* [x] Applies to #5209
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Validation Steps Performed
- Ran the tests.

Commit:
compare config files

Commit:
create settings files

Commit:
updated paths

Commit:
updated paths

Commit:
updated paths

Commit:
updated path check

ryanbodrug-microsoft commented on src/core/Microsoft.PowerToys.Settings.UI.Lib/ViewModels/ColorPickerViewModel.cs:
```diff
@@ -113,10 +117,15 @@ public int CopiedColorRepresentationIndex
         {
             SendConfigMSG(
-                   string.Format("{{ \"powertoys\": {{ \"{0}\": {1} }} }}", ColorPickerSettings.ModuleName, JsonSerializer.Serialize(_colorPickerSettings)));
+                   string.Format("{{ \"powertoys\": {{ \"{0}\": {1} }} }}", GetSettingsSubPath(), JsonSerializer.Serialize(_colorPickerSettings)));
```
Doesn't this break IPC communication?  Are you sure color picker changes are being saved?

laviusmotileng-ms replied:
i reverted the changes

ryanbodrug-microsoft commented:
@laviusmotileng-ms  / @crutkas .   Maybe I misunderstood the intent of #5209.  I thought we wanted to make sure that we are still able to Deserialize save files from older versions of powertoys.  Something like. 

1. Install previous versions of powertoys.  
2. For each module:
 2.1. Change settings from defaults. 
 2.2  Save serialized savefile. 
3.  For each module in each previous version. 
3.1 Load the serialized save file as an embedded resource or hard coded string, if small enough. 
3.2 Write a unit test that loads the file and verifies the expected result. 

It looks to me that your just verifying the default values of the serialized save data.   Am I missing something?

Commit:
reverted color picker changes

laviusmotileng-ms commented:
> @laviusmotileng-ms / @crutkas . Maybe I misunderstood the intent of #5209. I thought we wanted to make sure that we are still able to Deserialize save files from older versions of powertoys. Something like.
> 
> 1. Install previous versions of powertoys.
> 2. For each module:
>    2.1. Change settings from defaults.
>    2.2  Save serialized savefile.
> 3. For each module in each previous version.
>    3.1 Load the serialized save file as an embedded resource or hard coded string, if small enough.
>    3.2 Write a unit test that loads the file and verifies the expected result.
> 
> It looks to me that your just verifying the default values of the serialized save data. Am I missing something?

my understanding was that we need to verify that old settings are not overwritten by the changes in current version upon installation of the new version (which would be current version).

Commit:
added test files

Commit:
removed settings uralted files

laviusmotileng-ms commented:
@ryanbodrug-microsoft I added test files to the project.

crutkas commented:
@ryanbodrug-microsoft what needs to happen here to get this in a good state for the test to be added in

ryanbodrug-microsoft committed:
Merging w/ latest from main and replacing mocks

ryanbodrug-microsoft committed:
Merge branch 'master' into user/lamotile/backwardCompatiblityTests

ryanbodrug-microsoft committed:
Merging latest, and getting it to compile

ryanbodrug-microsoft commented:
didn't mean to push to this branch. Marking as draft as these tests don't do anything right now.

ryanbodrug-microsoft committed:
Modifying directory structure and adding properties for backcompat test.

ryanbodrug-microsoft committed:
Updating Unit test to use mock repositories.
1) BackCompatTestProperties exposes mock repository implementation because SettingsRepository is a singleton, and settings utils isn't typed.
2) BackCompatTestProperties, encapsulates logic to verify that the correct file was read from.
3) Validating each file is read twice.  Once by the original file, and once via the view model.

ryanbodrug-microsoft committed:
Adding 18.2 settings files.

ryanbodrug-microsoft committed:
Merge branch 'master' into user/lamotile/backwardCompatiblityTests

ryanbodrug-microsoft committed:
Fix compiler errors from latest merge

ryanbodrug-microsoft committed:
Adding v0.19.2 test files.

ryanbodrug-microsoft committed:
Adding in 0.20.1 settings.  Removing 0.20.2 (as this was a dev build number)

ryanbodrug-microsoft committed:
Adding settings tests for 22.1

ryanbodrug-microsoft committed:
General Settings should update version when they don't match

ryanbodrug-microsoft committed:
Adding v0.22.0 files

ryanbodrug-microsoft committed:
Removing not settings related files from TestData for PT Run 21.1

alekhyareddy28 commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/BackwardsCompatibility/BackCompatTestProperties.cs:
```diff
@@ -0,0 +1,77 @@
+    {
+        public const string RootPathStubFiles = "..\\..\\..\\..\\src\\core\\Microsoft.PowerToys.Settings.UI.UnitTests\\BackwardsCompatibility\\TestFiles\\{0}\\Microsoft\\PowerToys\\{1}\\{2}";
+
+        internal class MockSettingsRepository<T> : ISettingsRepository<T> where T : ISettingsConfig, new()
+        {
+            T _settingsConfig;
+            ISettingsUtils _settingsUtils;
+            public MockSettingsRepository( ISettingsUtils settingsUtils)
+            {
+                _settingsUtils = settingsUtils;
+            }
```
I might be missing something but MockSettingsRepository is not different from Settings Repository and it's also no longer a singleton. Why was this needed?

ryanbodrug-microsoft replied:
That _is_ the difference.  One of the drawbacks of singletons is that they are difficult to use in unit tests.  There are two reasons this needs to be mocked. 
1) I want to verify that the mocked io provider is read for a particular settings file.  If i use the singleton then then I can't verify the number of times the Read method is called. 
2) While the settings repository is typed, the ISettingsUtils and IOProvider aren't.  This means that the settings repository as a singleton will return the first instance of SettingsUtils that it is created with.    We may want to reconsider the GetInstance implementation in that aspect, but that was oustide the scope of this PR IMHO.

alekhyareddy28 replied:
Thanks for the clarification. Regarding 2, I agree that we have the issue where SettingsRepository would return the SettingsUtils of the type that it is created with.

alekhyareddy28 commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/PowerLauncherViewModelTest.cs:
```diff
@@ -5,13 +5,21 @@
     {
+        // This should not be changed. 
+        // Changing it will causes user's to lose their local settings configs.
+        public const string OriginalModuleName = "PowerToys Run";
```
I think for all modules other than ImageResizer we can use <Module>Settings.ModuleName to prevent redeclaration.

alekhyareddy28 commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/ColorPicker.cs:
```diff
@@ -2,13 +2,73 @@
+        /// <summary>
+        /// Test if the original settings files were modified.
+        /// </summary>
+        [TestMethod]
+        [DataRow("v0.20.1", "settings.json")]  //Color picker was introduced in .20
+        [DataRow("v0.21.1", "settings.json")]
+        [DataRow("v0.22.0", "settings.json")]
+        public void OriginalFilesModificationTest(string version, string fileName)
+        {
+            //Arrange
+            var mockIOProvider = BackCompatTestProperties.GetModuleIOProvider(version, OriginalModuleName, fileName);
+            var mockSettingsUtils = new SettingsUtils(mockIOProvider.Object);
+            ColorPickerSettings originalSettings = mockSettingsUtils.GetSettings<ColorPickerSettings>(OriginalModuleName);
+
+            var mockGeneralIOProvider = BackCompatTestProperties.GetGeneralSettingsIOProvider(version);
+            var mockGeneralSettingsUtils = new SettingsUtils(mockGeneralIOProvider.Object);
+            GeneralSettings originalGeneralSettings = mockGeneralSettingsUtils.GetSettings<GeneralSettings>();
+            var generalSettingsRepository = new BackCompatTestProperties.MockSettingsRepository<GeneralSettings>(mockGeneralSettingsUtils);
+
+            //Act
+            // Initialise View Model with test Config files
+            ColorPickerViewModel viewModel = new ColorPickerViewModel(mockSettingsUtils, generalSettingsRepository, ColorPickerIsEnabledByDefault_IPC);
+
+            //Assert
+            // Verifiy that the old settings persisted
```
I think I'm missing the idea behind the tests. Was it to validate that configurations persist everytime we open powertoys of a specific version or that the configurations of the powertoys persist on version upgrade? The tests seem to be doing the former but shouldn't we also check the later?

ryanbodrug-microsoft replied:
> I think I'm missing the idea behind the tests. Was it to validate that configurations persist everytime we open powertoys of a specific version or that the configurations of the powertoys persist on version upgrade? The tests seem to be doing the former but shouldn't we also check the later?

The idea was to validate the scenario that the the current code can handle loading any previous save file.  We don't want to assume which version they are upgrading from.   If we support a setting it should be exposed via the view model, if we don't then then it shouldn't crash or prevent any of the settings that exist across pt versions from loading.

alekhyareddy28 commented on src/core/Microsoft.PowerToys.Settings.UI.Lib/GeneralSettings.cs:
```diff
@@ -98,7 +98,7 @@ public bool UpgradeSettingsConfiguration()
             try
             {
-                if (Helper.CompareVersions(PowertoysVersion, Helper.GetProductVersion()) < 0)
+                if (Helper.CompareVersions(PowertoysVersion, Helper.GetProductVersion()) != 0)
```
It's nice to have this so that the settings file is of the correct version if someone chooses to downgrade. However, do we want to add tests so that everything works as expected when we downgrade?

ryanbodrug-microsoft replied:
I don't think that is possible with a unit test.   This change wasn't to support downgrading becuase the code being tested will always be the latest.  However if we have experimental releases in the wild this might be necessary as we This change was necessary because at the time of writing we released an experimental .22 release, but the current product version was .21.  This would force the tests that validated that we serialized the current version to fail.   We could either remove the .22 data from this test as it's experimental or leave it as is.     I personally thinks it makes sense as is because we want to support the .22 -> .23  path as well.  Plus once we have dev product versions in the pipeline we will still probably want this logic.

alekhyareddy28 reviewed (commented):
Validated that all the tests pass. 

Are we planning to add tests for compatibility between the old and new settings as well?

ryanbodrug-microsoft commented:
> Validated that all the tests pass.
> 
> Are we planning to add tests for compatibility between the old and new settings as well?

Can you be more specific here?  That is the intent of these tests.  Did I miss something?

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/FancyZones.cs:
```diff
@@ -19,6 +21,70 @@ namespace ViewModelTests
+            BackCompatTestProperties.VerifyGeneralSettingsIOProviderWasRead(mockGeneralIOProvider, expectedCallCount);
+        }
+
+        private int ColorPickerIsEnabledByDefault_IPC(string msg)
+        {
+            OutGoingGeneralSettings snd = JsonSerializer.Deserialize<OutGoingGeneralSettings>(msg);
+            Assert.IsTrue(snd.GeneralSettings.Enabled.ColorPicker);
+            return 0;
+        }
```
This seems to have been added accidentally?

ryanbodrug-microsoft replied:
> This seems to have been added accidentally?

Maybe?  I thought it was in one of lavius's earlier changes, but i'll double check.

alekhyareddy28 replied:
I think this code was initially removed and was added in the GeneralSettings test to verify that all the modules are enabled by default.

arjunbalgovind commented on src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/General.cs:
```diff
@@ -21,11 +23,56 @@ public class General
+            Func<string, int> SendMockIPCConfigMSG = msg => { return 0; };
+            Func<string, int> SendRestartAdminIPCMessage = msg => { return 0; };
+            Func<string, int> SendCheckForUpdatesIPCMessage = msg => { return 0; };
+            GeneralViewModel viewModel = new GeneralViewModel(
+                generalSettingsRepository,
+                "GeneralSettings_RunningAsAdminText",
+                "GeneralSettings_RunningAsUserText",
+                false,
+                false,
+                UpdateUIThemeMethod,
+                SendMockIPCConfigMSG,
+                SendRestartAdminIPCMessage,
+                SendCheckForUpdatesIPCMessage,
+                string.Empty);
```
[nit] Can we use named args for this particular declaration since there are many arguments?

alekhyareddy28 commented:
> > Validated that all the tests pass.
> > Are we planning to add tests for compatibility between the old and new settings as well?
> 
> Can you be more specific here? That is the intent of these tests. Did I miss something?

Maybe I missed it but did we use the settings files generated by the older settings process? If so please ignore the rest of this.

I was referring to this scenario with image resizer. So someone might initially be using the older C++ settings and then when they upgrade their build and use the C# settings does everything work as expected. I was especially coming from the Image Resizer side where some parts have been moved to the settings from the Img Resizer code. I'm not completely sure how (the format) and where the older settings saved the information for Img Resizer. For most of the powertoys the config is saved using 'set_config' so the format should be the same irrespective of the old or new settings but ImgResizer doesn't seem to do that (https://github.com/microsoft/PowerToys/blob/master/src/modules/imageresizer/dll/dllmain.cpp#L75). The new settings process (ImageResizerViewModel.cs) saves the configurations directly without passing them to runner because some options were moved to it, so I was thinking it would be a good check instead of maybe creating a new configurations file if there is a format exception.

ryanbodrug-microsoft committed:
Referencing module name as *Settings.ModuleName in tests.  Except for ImageResizer

ryanbodrug-microsoft committed:
Using ImageResizerSettings to use Settings for the module name.

ryanbodrug-microsoft committed:
Setting AllPlugins to empty list in case PluginManager.Save/Load/ReloadData is called before plugins are loaded

ryanbodrug-microsoft committed:
Merge branch 'master' of https://github.com/microsoft/PowerToys

alekhyareddy28 commented:
@ryanbodrug-microsoft, there are two places where the name of a module is used - 
1. As the folder name while being stored ([here](https://github.com/microsoft/PowerToys/blob/master/src/core/Microsoft.PowerToys.Settings.UI.Lib/ImageResizerSettings.cs#L13) for ImageResizer)
2. As information being sent to the runner - ([this](https://github.com/microsoft/PowerToys/blob/master/src/runner/general_settings.cpp#L184) is where the name must match the name of the module to enable or disable the PowerToy.

For most of the powerToys both these values are the same. This does not seem to be the case for ImageResizer. The ImageResizer folder name is `ImageResizer` without the space. This would work with imageresizer while using the older settings as well because the name accessed directly by imagesizer previously is also without the space ([here](https://github.com/microsoft/PowerToys/blob/master/src/modules/imageresizer/dll/Settings.cpp#L47)).

However while sending information to the runner, the name of the module must be "Image Resizer"(With the space) and this is based on the general settings enabled modules config [here](https://github.com/microsoft/PowerToys/blob/master/src/core/Microsoft.PowerToys.Settings.UI.Lib/EnabledModules.cs#L37). 
That being said the enable/disable works fine with the current changes however the only change is the name of the module in the settings.json file from 'Image Resizer' to 'ImageResizer'. I'm not sure how that would affect the older settings, will look into that and let you know.

ryanbodrug-microsoft commented:
> @ryanbodrug-microsoft, there are two places where the name of a module is used -
> 
> 1. As the folder name while being stored ([here](https://github.com/microsoft/PowerToys/blob/master/src/core/Microsoft.PowerToys.Settings.UI.Lib/ImageResizerSettings.cs#L13) for ImageResizer)
> 2. As information being sent to the runner - ([this](https://github.com/microsoft/PowerToys/blob/master/src/runner/general_settings.cpp#L184) is where the name must match the name of the module to enable or disable the PowerToy.
> 
> For most of the powerToys both these values are the same. This does not seem to be the case for ImageResizer. The ImageResizer settings storage folder name is `ImageResizer` without the space. This would work with the older settings as well because the name accessed directly by imagesizer previously is also without the space ([here](https://github.com/microsoft/PowerToys/blob/master/src/modules/imageresizer/dll/Settings.cpp#L47)).
> 
> However while sending information to the runner, the name of the module must be "Image Resizer"(With the space) and this is based on the names of the powertoys according to the runner and is followed by the general settings enabled modules config [here](https://github.com/microsoft/PowerToys/blob/master/src/core/Microsoft.PowerToys.Settings.UI.Lib/EnabledModules.cs#L37).
> That being said the enable/disable works fine with the current changes however the only change is the name of the module in the settings.json file from 'Image Resizer' to 'ImageResizer'. I'm not sure how that would affect the older settings, will look into that and let you know.

Just to confirm.  The link you pointed to for 1. doesn't actually provide the folder name (before this change).  That is provided by the module name in ImageResizerViewModel.cs.   I believe the link you provided is only displayed internally in the file, and to my knowledge that is not used anywwhere.     I _think_ we are saying the same thing, but just wanted to confirm.

alekhyareddy28 reviewed (approved):
@ryanbodrug-microsoft, yes we're both referring to the same thing. Sorry if I wasn't clear. 

Regarding the change in the module name in the settings.json, seems like the older settings.json file also had `ImageResizer` (without the space) ([here](https://github.com/microsoft/PowerToys/blob/master/src/modules/imageresizer/ui/Properties/Settings.cs#L370)) so with the changes in this PR we are infact making it uniform. Also, as you mentioned I don't think anything is actually reading that value. 

Other than the couple of minor comments that Arjun made, the PR LGTM!

ryanbodrug-microsoft committed:
Merge branch 'master' of https://github.com/microsoft/PowerToys

ryanbodrug-microsoft committed:
Merging latest from master, resolving conflicts with imageresizer tests

ryanbodrug-microsoft committed:
Fixing fxcop errors

ryanbodrug-microsoft committed:
using named parameters as per review feedback

crutkas commented:
/azp run
